keps/sig-cluster-lifecycle/wgs/20200108-instance-specific-componentconfig.md

```diff
@@ -0,0 +1,644 @@
+verbose. In the future, we can explore more elegant ways of generating files
+with instance-specific parameters correctly substituted. This proposal argues
+that the extra verbosity is worth the benefit of a consistent, versioned API for
+configuring core components.
+
+## Design Details
+
+The canonical flag for providing a ComponentConfig file to a component is
+`--config`. This KEP proposes that the canonical flag for providing an
+instance-specific config file be `--instance-config`.
+
+As with sharable ComponentConfig parameters, command line flags for
+instance-specific parameters should continue to function and take precedence
+over the final merged values from the config file so that backward compatibility
+is maintained.
+
+### Configuration merge example
+
+The merge between both YAML configuration files is provided by the
+[Strategic Merge Patch](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/strategic-merge-patch.md)
+library.
+
+The example below is a sharable Kubelet configuration YAML file.
+
+```yaml
+apiVersion: kubelet.config.k8s.io/v1beta1
+authorization:
+  mode: Webhook
+  webhook:
+    cacheAuthorizedTTL: 0s
+    cacheUnauthorizedTTL: 0s
+clusterDNS:
+- 10.96.0.10
+clusterDomain: cluster.local
+cpuManagerReconcilePeriod: 0s
+evictionHard:
+  imagefs.available: 0%
+  nodefs.available: 0%
+  nodefs.inodesFree: 0%
+evictionPressureTransitionPeriod: 0s
+featureGates:
+  DynamicKubeletConfig: true
+healthzBindAddress: 127.0.0.1
+healthzPort: 10248
+kind: KubeletConfiguration
+rotateCertificates: true
+staticPodPath: /etc/kubernetes/manifests
+```
+
+This is the instance-configuration YAML, which is going to be merged into the sharable:
+
+```yaml
+address: 127.0.0.1
+apiVersion: kubelet.config.k8s.io/v1beta1
+clusterDNS:
+- 10.96.0.11
+evictionHard:
+  imagefs.available: 2%
+featureGates:
+  TaintBasedEvictions: true
+kind: KubeletConfiguration
+```
+
+The final patched YAML, used to unmarshal into the KubeletConfiguration object is:
+
+```yaml
+address: 127.0.0.1
+apiVersion: kubelet.config.k8s.io/v1beta1
+authorization:
+  mode: Webhook
+  webhook:
+    cacheAuthorizedTTL: 0s
+    cacheUnauthorizedTTL: 0s
+clusterDNS:
+- 10.96.0.11
+clusterDomain: cluster.local
+cpuManagerReconcilePeriod: 0s
+evictionHard:
+  imagefs.available: 2%
+  nodefs.available: 0%
+  nodefs.inodesFree: 0%
+evictionPressureTransitionPeriod: 0s
+featureGates:
+  DynamicKubeletConfig: true
+  TaintBasedEviction: true
+healthzBindAddress: 127.0.0.1
+healthzPort: 10248
+kind: KubeletConfiguration
+rotateCertificates: true
+staticPodPath: /etc/kubernetes/manifests
+```
+
+The approach of performing a strategic merge patch of instance-specific over
+shareable config before unmarshalling can decrease the complexity by:
+
+- Using an already proven method from apimachinery.
+- Avoiding complexity around automatic conversions during unmarshaling that might
```

Comment:
orthogonal: we should strive to default as little as possible, and fix some of the API machinery code that always defaults to defaulting when decoding. See my out-of-tree proposed fix for this: https://pkg.go.dev/github.com/weaveworks/libgitops/pkg/serializer